File: iqchannels-sdk/src/main/java/ru/iqchannels/sdk/schema/ChatMessage.java
package ru.iqchannels.sdk.schema;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.util.Date;
import java.util.List;

import ru.iqchannels.sdk.http.HttpRequest;

/**
 * Copyright iqstore.ru. All Rights Reserved.
 */
public class ChatMessage {
    public long Id;
    @Nullable public String UID;
    public long ChatId;
    public long SessionId;
    public long LocalId;
    @Nullable public Long EventId;
    public boolean Public;

    // Author
    public ActorType Author;
    @Nullable public Long ClientId;
    @Nullable public Long UserId;

    // Payload
    public String Payload; // ChatPayloadType
    public String Text;
    @Nullable public String FileId;
    @Nullable public String NoticeId;
    @Nullable public Long RatingId;

    @Nullable public Long ReplyToMessageId;

    @Nullable public Boolean IsDropDown;
    @Nullable public Boolean DisableFreeText;
    @Nullable public List<SingleChoice> SingleChoices;
    @Nullable public List<Action> Actions;

    // Flags
    public boolean Received;
    public boolean Read;

    public long CreatedAt;
    @Nullable public Long ReceivedAt;
    @Nullable public Long ReadAt;

    // Transient
    public boolean My;

    // Local
    public boolean Sending;
    @Nullable public Client Client;
    @Nullable public User User;
    @Nullable public UploadedFile File;
    @Nullable public Rating Rating;
    @Nullable public Date Date;

    // Upload
    @Nullable public transient File Upload;
    @Nullable public transient Exception UploadExc;
    @Nullable public transient HttpRequest UploadRequest;
    public transient int UploadProgress;

    public ChatMessage() {}

    public ChatMessage(@NonNull Client client, long localId, String text) {
        this(client, localId);

        Payload = ChatPayloadType.TEXT;
        Text = text;
    }

    public ChatMessage(@NonNull Client client, long localId, @NonNull File file, @Nullable Long replyToMessageId) {
        this(client, localId);

        Payload = ChatPayloadType.FILE;
        Text = file.getName();
        Upload = file;
        ReplyToMessageId = replyToMessageId;
    }

    public ChatMessage(@NonNull Client client, long localId) {
        LocalId = localId;
        Public = true;

        // Author
        Author = ActorType.CLIENT;
        ClientId = client.Id;

        Date now = new Date();
        CreatedAt = now.getTime();

        // Transitive
        My = true;

        // Local
        Client = client;
        Date = now;
    }

    public ChatMessage(@NonNull User user, long localId) {
        LocalId = localId;
        Public = true;

        // Author
        Author = ActorType.USER;
        UserId = user.Id;

        Date now = new Date();
        CreatedAt = now.getTime();

        // Transitive
        My = false;

        // Local
        User = user;
        Date = now;
    }
}

File: iqchannels-sdk/src/main/java/ru/iqchannels/sdk/schema/ChatMessage.kt
package ru.iqchannels.sdk.schema

import com.google.gson.Gson
import java.io.File
import java.util.*
import ru.iqchannels.sdk.http.HttpRequest

class ChatMessage {
	var Id: Long = 0
	var UID: String? = null
	var ChatId: Long = 0
	var SessionId: Long = 0
	var LocalId: Long = 0
	var EventId: Long? = null
	var Public = false

	// Author
	var Author: ActorType? = null
	var ClientId: Long? = null
	var UserId: Long? = null

	// Payload
	var Payload: String? = null // ChatPayloadType
	var Text: String? = null
	var FileId: String? = null
	var NoticeId: String? = null
	var RatingId: Long? = null
	var ReplyToMessageId: Long? = null
	var IsDropDown: Boolean? = null
	var DisableFreeText: Boolean? = null
	var SingleChoices: List<SingleChoice>? = null
	var Actions: List<Action>? = null

	// Flags
	var Received = false
	var Read = false
	var CreatedAt: Long = 0
	var ReceivedAt: Long? = null
	var ReadAt: Long? = null

	// Transient
	var My = false

	// Local
	var Sending = false
	var System = false
	var Client: Client? = null
	var User: User? = null
	var File: UploadedFile? = null
	var Rating: Rating? = null
	var Date: Date? = null
	var newMsgHeader: Boolean = false

	override fun toString(): String {
		return Gson().toJson(this)
	}

	// Upload
	@Transient
	var Upload: File? = null

	@Transient
	var UploadExc: Exception? = null

	@Transient
	var UploadRequest: HttpRequest? = null

	@Transient
	var UploadProgress = 0

	constructor()
	constructor(client: Client, localId: Long, text: String?) : this(client, localId) {
		Payload = ChatPayloadType.TEXT
		Text = text
	}

	constructor(client: Client, localId: Long, text: String, file: File, replyToMessageId: Long?) : this(
		client,
		localId
	) {
		Payload = ChatPayloadType.FILE
		Text = text
		Upload = file
		ReplyToMessageId = replyToMessageId
	}

	constructor(client: Client, localId: Long) {
		LocalId = localId
		Public = true

		// Author
		Author = ActorType.CLIENT
		ClientId = client.Id
		val now = Date()
		CreatedAt = now.time

		// Transitive
		My = true

		// Local
		Client = client
		Date = now
	}

	constructor(user: User, localId: Long) {
		LocalId = localId
		Public = true

		// Author
		Author = ActorType.USER
		UserId = user.Id
		val now = Date()
		CreatedAt = now.time

		// Transitive
		My = false

		// Local
		User = user
		Date = now
	}
}
